Userland/Libraries/LibPDF/ColorSpace.cpp

```diff
@@ -165,24 +165,58 @@ PDFErrorOr<NonnullRefPtr<DeviceNColorSpace>> DeviceNColorSpace::create(Document*
 
     // "The names parameter is an array of name objects specifying the individual color components.
     //  The length of the array determines the number of components in the DeviceN color space"
-    auto names = parameters[0].get<NonnullRefPtr<Object>>()->cast<ArrayObject>();
+    auto names_array = TRY(document->resolve_to<ArrayObject>(parameters[0]));
+    Vector<DeprecatedString> names;
+    for (size_t i = 0; i < names_array->size(); ++i)
+        names.append(names_array->get_name_at(i)->name());
 
     // "The alternateSpace parameter is an array or name object that can be any device or CIE-based color space
     //  but not another special color space (Pattern, Indexed, Separation, or DeviceN)."
+    auto alternate_space_object = TRY(document->resolve_to<Object>(parameters[1]));
+    auto alternate_space = TRY(ColorSpace::create(document, alternate_space_object));
+
+    auto family = alternate_space->family();
+    if (family == ColorSpaceFamily::Pattern || family == ColorSpaceFamily::Indexed || family == ColorSpaceFamily::Separation || family == ColorSpaceFamily::DeviceN)
+        return Error { Error::Type::MalformedPDF, "DeviceN color space has invalid alternate color space" };
+
+    // "The tintTransform parameter specifies a function"
+    auto tint_transform_object = TRY(document->resolve_to<Object>(parameters[2]));
+    auto tint_transform = TRY(Function::create(document, tint_transform_object));
 
-    // FIXME: Implement.
+    // FIXME: If `attributes` is present and has /Subtype set to /NChannel, possibly
+    //        do slightly different processing.
 
-    return adopt_ref(*new DeviceNColorSpace(names->size()));
+    auto color_space = adopt_ref(*new DeviceNColorSpace(move(alternate_space), move(tint_transform)));
+    color_space->m_names = move(names);
```

Comment:
I'm guessing there probably is, but is there a reason `names` isn't passed in as another constructor parameter?

(Code looks fine to my non-PDF-coloured eyes otherwise.)

Reply:
I'm passing in the ones I have to pass in (nonnullrefptrs) and setting the others directly -- it's a static method on the class itself after all.

(Less repetition in the ctor this way.)